### DISCUSSION TITLE
Better Integration with ASP.net Authorization middleware

### DISCUSSION THREAD

**Discussion by DanielLaberge** [CoreWCF/CoreWCF#783] [Ideas]
Would it be possible for CoreWCF to expose its Endpoints and Metadata to asp.net, so that middleware can access it through `HttpContext.GetEndpoint()`? 

This could allow asp.net's Authorization middleware and attributes like `[AllowAnonymous]` and `[Authorize]` to work the same way they do with MVC and WebApis

The app.[UseAuthorization() middleware uses metadata from the Endpoint](https://github.com/dotnet/aspnetcore/blob/main/src/Security/Authorization/Policy/src/AuthorizationMiddleware.cs#L39) to determine which Policies to apply to the request. Eg.: it checks for the `[AllowAnonymous]` attribute like so: `HttpContext.GetEndpoint()?.Metadata.GetMetadata<IAllowAnonymous>()`

This would be ideal and would avoid the headache of having to deal with IServiceBehavior, IOperationBehavior and IOperationInvoker, to implement attribute based authorization, which is how we did it in WCF.

**Comment by DanielLaberge** [comment]
In the meantime, I'm looking for a simplest way to require a valid user for all endpoints except the ones marked with [AllowAnonymous].

My goal would be to ideally avoid the WCF-specific Auth and Authorization features in favour of ASP.net's:
 
I'm using this fallback policy to require a valid user for everything, which works as expected:
``` 
        services.AddAuthentication("Bearer").AddJwtBearer(/*...*/);
        services.AddAuthorization(options =>
        {
            options.FallbackPolicy = new Microsoft.AspNetCore.Authorization.AuthorizationPolicyBuilder()
                .RequireAuthenticatedUser()
                .Build();
        });
      // ...
      app.UseAuthentication();
      app.UseAuthorization(); 
``` 

But I've yet to figure out a way to override it for OperationContracts marked with [AllowAnonymous], or for the wsdl endpoint. 
Any ideas?

P.S. I'm aware I could use Path based policies like showcased in #689, but I'm looking for something more flexible and robust based on attributes.

**Comment by mconnew** [comment]
I'm working with @samsp-msft to come up with some ways to achieve this. I think there are 3 pieces to the solution here. 
1. Making ClientCredentialType value of InheritedFromHost pick up the identity from HttpContext.User.Identity and populate it into the ServiceSecurityContext.  
2. Add support for using the asp.net core authentication mechanisms to be used with TransportWithMessageCredentials. This brings in support for other bindings such as NetTcp to easily use whatever mechanisms asp.net core has now or in the future. This allows any asp.net core authentication mechanism you like to be used with HTTP.  One of the goals of CoreWCF is "Own less code", which means we should leverage asp.net core functionality whenever we can.  
3. Add support to attribute operations to use asp.net core authorization policies.  

I haven't got deep enough into policies yet, but as far as I have got, they seem quite flexible. For the WSDL endpoint, could you have a different policy based on HTTP method? wsdl is fetched using a GET request, but regular operation calls are done using POST.  

Using the existing asp.net core attributes isn't a great option as we generally need to have attributes which implement IOperationBehavior to be automatically hooked in to change behavior. Would attributing your implementation methods with CoreWCF attributes which have matching capabilities be sufficient? Another option I can think of would be a service level attribute (which implements IServiceBehavior) which can reflect on the service type for the asp.net core attributes. I'm not sure how difficult that second option would be.

**Reply by samsp-msft** [reply]
I am hoping that we can get the WCF AuthorizeRole attribute hooked into the ASP.NET system - and we are working through that right now. The sample PR is at https://github.com/CoreWCF/samples/pull/21

As a temporary workaround (hopefully the above will only be a couple of weeks away), you can use imperative verification within the service methods themselves.
- Make the service class partial
- Register it with dependency injection using builder.Services.AddSingleton<typename>();
- Add an injected parameter for the HttpContext such as `public IList<GameTile> DrawTiles(int count, [Injected] HttpContext ctx)` , this signature is then different from the one in the contract, but codegen will be done to create that and hook it all up for you.
- within the service method, look at the context and do your auth there. for example MSAL for AAD has a bunch of extension methods to do just that based on the http context.

**Reply by DanielLaberge** [reply]
>Would attributing your implementation methods with CoreWCF attributes which have matching capabilities be sufficient?

Yes, that would be perfectly acceptable.

Thanks for the update.

**Reply by DanielLaberge** [reply]
@samsp-msft Where can we track the progression of adding support for AuthorizeRole? Or shall we reword this issue to serve this purpose?

Thank you.

**Reply by samsp-msft** [reply]
This probably needs to change from using CoreWCFs AuthorizeRole, to using ASP.NET's own Authorize attributes as role based authorization is not the preferred method for ASP.NET's authorization - that's to define claims and policies.

Sam

**Comment by DanielLaberge** [comment]
Agreed, using ASP.net’s attributes would be ideal. My question remains; is there an existing issue I can track so I can be notified when it’s ready to use?

**Reply by g7ed6e** [reply]
you can subscribe to #850
